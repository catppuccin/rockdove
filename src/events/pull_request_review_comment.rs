use octocrab::models::webhook_events::{
    WebhookEvent,
    payload::{
        PullRequestReviewCommentWebhookEventAction, PullRequestReviewCommentWebhookEventPayload,
    },
};

use crate::{
    colors::PULL_REQUEST_COLOR,
    embed_builder::EmbedBuilder,
    errors::{RockdoveError, RockdoveResult},
};

pub fn make_embed(
    event: WebhookEvent,
    specifics: &PullRequestReviewCommentWebhookEventPayload,
) -> RockdoveResult<Option<EmbedBuilder>> {
    if !matches!(
        specifics.action,
        PullRequestReviewCommentWebhookEventAction::Created
    ) {
        return Ok(None);
    }

    let repo = event
        .repository
        .ok_or_else(|| RockdoveError::MissingField {
            event_type: event.kind.clone(),
            field: "repository",
        })?;

    let mut embed = EmbedBuilder::default();

    let repo_name = repo.full_name.unwrap_or(repo.name);

    embed.title(&format!(
        "[{}] New review comment on pull request #{}: {}",
        repo_name,
        specifics.pull_request.number,
        specifics
            .pull_request
            .title
            .as_ref()
            .ok_or_else(|| RockdoveError::MissingField {
                event_type: event.kind.clone(),
                field: "pull_request.title",
            })?,
    ));
    embed.url(specifics.comment.html_url.as_str());
    embed.description(specifics.comment.body.as_str());
    embed.color(PULL_REQUEST_COLOR);

    Ok(Some(embed))
}

#[cfg(test)]
mod tests {
    use crate::snapshot_test;

    #[test]
    fn created() {
        snapshot_test!("pull_request_review_comment", "created");
    }
}
