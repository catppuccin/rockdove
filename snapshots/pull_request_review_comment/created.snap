---
source: src/events/pull_request_review_comment.rs
info:
  author_name_length: 8
  colour_hex: "#89B4FA"
  description_length: 70
  title_length: 106
---
embeds:
  - author:
      icon_url: "https://avatars.githubusercontent.com/u/47499684?v=4"
      name: uncenter
      url: "https://github.com/uncenter"
    color: 9024762
    description: "More clarity about what you are clicking on, why it would install it. "
    title: "[catppuccin/discord] New review comment on pull request #325: docs(README): update and improve usage steps"
    url: "https://github.com/catppuccin/discord/pull/325#discussion_r1785288051"
